shell/common/api/electron_api_asar.cc

```diff
@@ -103,15 +107,67 @@ class Archive : public gin_helper::Wrappable<Archive> {
     return gin::ConvertToV8(isolate, new_path);
   }
 
-  // Return the file descriptor.
-  int GetFD() const {
-    if (!archive_)
-      return -1;
-    return archive_->GetFD();
+  v8::Local<v8::ArrayBuffer> ReadSync(gin_helper::ErrorThrower thrower,
+                                      uint64_t offset,
+                                      uint64_t length) {
+    base::CheckedNumeric<uint64_t> safe_offset(offset);
+    base::CheckedNumeric<uint64_t> safe_end = safe_offset + length;
+    if (!safe_end.IsValid() ||
+        safe_end.ValueOrDie() > archive_->file()->length()) {
+      thrower.ThrowError("Out of bounds read requested in ASAR");
+      return v8::Local<v8::ArrayBuffer>();
+    }
+    auto array_buffer = v8::ArrayBuffer::New(thrower.isolate(), length);
+    auto backing_store = array_buffer->GetBackingStore();
+    memcpy(backing_store->Data(), archive_->file()->data() + offset, length);
+    return array_buffer;
+  }
+
+  v8::Local<v8::Promise> Read(gin_helper::ErrorThrower thrower,
+                              uint64_t offset,
+                              uint64_t length) {
+    gin_helper::Promise<v8::Local<v8::ArrayBuffer>> promise(thrower.isolate());
+    v8::Local<v8::Promise> handle = promise.GetHandle();
+
+    base::CheckedNumeric<uint64_t> safe_offset(offset);
+    base::CheckedNumeric<uint64_t> safe_end = safe_offset + length;
+    if (!safe_end.IsValid() ||
+        safe_end.ValueOrDie() > archive_->file()->length()) {
+      thrower.ThrowError("Out of bounds read requested in ASAR");
```

Comment:
This error should be passed in promise otherwise it would not be captured user's callback.

Reply:
Hm, interesting point. When using `async/await` either are captured by `try/catch`, the only difference is when using `.catch()` (or the 2-arg `.then()`). It would also make a difference if not `await`ing the result of such a promise.

e.g.


```js
throwsErrorsSynchronously(badArguments)
  .then(() => { /* ... */ })
  .catch((err) => { /* never reached */ })
// never reached:
doOtherThings().then(/* ... */)
```

vs. throwing errors by rejecting the promise:

```js
throwsErrorsAsynchronously(badArguments)
  .then(() => { /* ... */ })
  .catch((err) => { /* error is thrown here */ })
doOtherThings().then(/* ... this will end up happening ... */)
```


It looks like when you write an `async function`, errors are always thrown by rejecting the promise rather than synchronously thrown, even if the error happens synchronously:


```js
async function errorThrower() { throw new Error('no anchovies unless otherwise specified') }
errorThrower()
  .catch((err) => { /* error appears here */ })
thisDoesGetCalled()
```

So I'll go with the way you suggest.